include/wil/com.h

```diff
@@ -3002,6 +3002,155 @@ namespace wil
+    {
+        using Interface = Itf;
+        using Result = T;
+    };
+
+    template<typename T>
+    struct has_next
+    {
+        template <typename U = T>
+        static auto test(int) -> decltype(wistd::declval<U>()->Next(0, nullptr, nullptr), wistd::true_type{});
+
+        template <typename>
+        static auto test(...) -> wistd::false_type;
+
+        static constexpr bool value = decltype(test<T>(0))::value;
+    };
+
+    template<typename T>
+    constexpr bool has_next_v = has_next<T>::value;
+
+    template<typename Interface>
+    struct com_enumerator_traits
+    {
+        using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
+        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, we use the raw pointer type IFoo*.
+        using smart_result = wistd::conditional_t<wistd::is_pointer_v<Result> && wistd::is_base_of_v<::IUnknown, wistd::remove_pointer_t<Result>>,
+            wil::com_ptr<wistd::remove_pointer_t<Result>>, Result>;
+    };
+}
+
+template <typename IEnumType, typename TStoredType = typename details::com_enumerator_traits<IEnumType>::smart_result>
+struct com_iterator
+{
+    wil::com_ptr<IEnumType> m_enum{};
+    TStoredType m_currentValue{};
+
+    com_iterator(com_iterator&&) = default;
+    com_iterator(com_iterator const&) = default;
+    com_iterator& operator=(com_iterator&&) = default;
+    com_iterator& operator=(com_iterator const&) = default;
+
+    com_iterator(IEnumType* enumPtr) : m_enum(enumPtr)
+    {
+        FetchNext();
+    }
+
+    auto operator->()
+    {
+        return wistd::addressof(m_currentValue);
+    }
+
+    auto& operator*()
+    {
+        return m_currentValue;
+    }
+
+    const auto& operator*() const
+    {
+        return m_currentValue;
+    }
+
+    com_iterator& operator++()
+    {
+        // If we're already at the end, don't try to advance. Otherwise, use Next to advance.
+        if (m_enum)
+        {
+            FetchNext();
+        }
+
+        return *this;
+    }
+
+    bool operator!=(com_iterator const& other) const
+    {
+        return !(*this == other);
+    }
+
+    bool operator==(com_iterator const& other) const
+    {
+        return (m_enum.get() == other.m_enum.get());
+    }
+
+private:
+    void FetchNext()
+    {
+        if (m_enum)
+        {
+            // we cannot say m_currentValue = {} because com_ptr has 2 operator= overloads: one for T* and one for nullptr_t
+            m_currentValue = TStoredType{};
+            auto hr = m_enum->Next(1, &m_currentValue, nullptr);
+            if (hr == S_FALSE)
+            {
+                m_enum = nullptr;
+            }
+            else 
+            {
+                THROW_IF_FAILED_MSG(hr, "Failed to get next");
+            }
+        }
+    }
+};
+
+
+template<typename IEnumXxx, wistd::enable_if_t<wil::details::has_next_v<IEnumXxx*>, int> = 0>

```

Comment:
We should allow the caller to specify the custom smart type to `make_range`. 

```cpp
for (auto&& pidl : make_range<IEnumIDList, wil::unique_pidl>(penum.get()))
{
   ...
}
```

Otherwise you have a memory leak because we leak all the pidls.

We should require a smart type if the enumerator produces a pointer.

```cpp
for (auto&& pidl : make_range(penum.get())) // should be illegal - leaks the pidls
{
   ...
}
```



Unfortunately, we cannot distinguish between returning structures with nothing worth destructing
```
struct SafeToJustAbandon
{
   int first, last;
};
```

and structures that require cleanup.

```cpp
struct MustCleanMeUpOrIWillLeak
{
    VARIANT v; // if you fail to VariantClear me, you will leak memory!
};
```

Reply:
@oldnewthing Originally this had the TStoredType with a default value of `smart_result`. However, @dunhor mentioned we should sfinae this method, so that requires an optional non-type template param.
So in order to support both I think you'd have to write `make_range<IEnumIDList, int, wil::unique_pidl>` (because putting the TStoredType first causes multiple definition IIRC), see https://github.com/microsoft/wil/pull/398#discussion_r1411281525

Reply:
_Technically_ the SFINAE part doesn't need to be a template parameter (though with an `auto` return type, the list of tricks available to you decreases). The original code also had a problem where an additional compilation error was coming in due to a non-dependent (i.e. not a substitution failure) access of the `T::Next` function pointer.

In Raymond's example, `TStoredType` is deduced "as necessary" and doesn't seem to have any of the original issues: https://godbolt.org/z/GTTaa6b7h

Reply:
Wearing my "samples-driven API design" hat, I figure that we want the calling code to look like this:
```cpp
wil::com_ptr<IEnumUnknown> enumUnknown;
wil::com_ptr<IEnumIDList> enumIDList;

// allowed: deduces IEnumUnknown and com_ptr<IUnknown>
auto enumerator = wil::com_enumerator(enumUnknown.get());

// allowed: deduces IEnumUnknown and uses caller-specified com_ptr_nothrow<IUnknown>
auto enumerator = wil::com_enumerator<wil::com_ptr_nothrow<IUnknown>>(enumUnknown.get());

// allowed: deduces IEnumIDList, uses explicitly-provided unique_idlist
auto enumerator = wil::com_enumerator<wil::unique_idlist>(enumIDLIst.get());

// not allowed - need to specify how to free the IDList
auto enumerator = wil::com_enumerator(enumIDLIst.get());
```
And similarly for `make_range`.

This means that we have
```cpp
template<typename TStoredType = void, typename IEnumXxx>
struct com_enumerator
{
    static_assert(!std::is_same_v<typename com_enumerator_traits::Result, typename com_enumerator_traits::smart_result> || !std::is_same_v<TStoredType, void>, "You must specify a type to receive the enumerated objects.")
};
```
possibly with SFINAE stuff added to fail the com_enumerator at deduction time rather than at instantiation.

Reply:
> possibly with SFINAE stuff added to fail the com_enumerator at deduction time rather than at instantiation.

If it's just a struct, there's no real need for SFINAE because there's no overload resolution for structs. It's _occasionally_ useful to SFINAE on specializations, but I can't imagine that being necessary here

Reply:
following up in #404